convert2rhel/utils.py

```diff
@@ -66,6 +69,141 @@ class Color(object):
 
+# Code taken from https://stackoverflow.com/a/63773140
+class Process(multiprocessing.Process):
+    """Overrides the implementation of the multiprocessing.Process class."""
+
+    def __init__(self, *args, **kwargs):
+        """Default constructor for the class"""
+        multiprocessing.Process.__init__(self, *args, **kwargs)
+        self._pconn, self._cconn = multiprocessing.Pipe()
+        self._exception = None
+
+    def run(self):
+        """Overrides the `run` method to catch exceptions raised in child."""
+        try:
+            multiprocessing.Process.run(self)
+            self._cconn.send(None)
+        # Here, `SystemExit` is inherit from `BaseException`, which is too
+        # broad to catch as it involves for-loop exceptions. The idea here
+        # is to catch `SystemExit` *and* any `Exception` that shows up as we do
+        # a lot of logger.critical() and they do raise `SystemExit`.
+        except (Exception, SystemExit) as e:
+            tb = traceback.format_exc()
+            self._cconn.send((e, tb))
+
+    @property
+    def exception(self):
+        """Custom property to access the exception raised by the children."""
+        if self._pconn.poll():
+            self._exception = self._pconn.recv()
+        return self._exception
+
+
+def run_as_child_process(func):
+    """Decorator to execute functions as child process.
+
+    This decorator will use `multiprocessing.Process` class to initiate the
+    function as a child process to the parent process with the intention to
+    execute that in its own process, thus, avoiding cases where libraries would
+    install handlers that could be propagated to the main thread if they do not
+    do a clean and proper clean-up.
+
+    .. note::
+        This decorator is mostly intended to be used when dealing with
+        functions that call directly `rpm` or `yum`.
+
+        In `rpm` (version 4.11.3, RHEL 7), it's know that whenever there is a
+        call to that library, it will install a global signal handler catching
+        different types of signals, but most importantly, the SIGINT (Ctrl + C)
+        one, which causes problem during the conversion, as we can't replace or
+        override that signal as it was initiated in a C library rather than a
+        python library.
+
+        By using this decorator, `rpm` will install the signal handler in the
+        child process and leave the parent one with the original signal
+        handling that is initiated by python itself (or, whatever signal is
+        registered when the conversion starts as well).
+
+    .. warning::
+        This decorator will use a instance of `multiprocessing.Queue` to
+        communicate between the processes, so, if this decorator is being used
+        in a function that should return values, it will need to adapt to the
+        queue workflow, and instead of returning it directly, just put the
+        value inside the queue and the decorator will pop it out once the
```

Comment:
Is the ` a return value` there correct? It should be `return value`, right?

Reply:
Ah, sorry, you're right. It should be `return a value`.

Reply:
Just removed that whole paragraph as it was about the old way that should adapt the functions to have a queue parameter, since it's changed, it is not needed anymore.